ot/mpz-ot-core/src/ferret/spcot/receiver.rs

```diff
@@ -0,0 +1,316 @@
+        }
+
+        let ExtendFromSender { ms, sum } = extendfs;
+        if ts.len() != h {
+            return Err(ReceiverError::InvalidLength(
+                "the length of t should be h".to_string(),
+            ));
+        }
+
+        if ms.len() != h {
+            return Err(ReceiverError::InvalidLength(
+                "the length of M should be h".to_string(),
+            ));
+        }
+
+        // Updates hasher
+        self.state.hasher.update(&ms.to_bytes());
+        self.state.hasher.update(&sum.to_bytes());
+
+        let mut alpha_bar_vec = alpha.to_msb0_vec();
+        alpha_bar_vec.drain(0..alpha_bar_vec.len() - h);
+
+        alpha_bar_vec.iter_mut().for_each(|a| *a = !*a);
+
+        // Setp 5 in Figure 6.
+        let k: Vec<Block> = ms
+            .into_iter()
+            .zip(ts)
+            .zip(alpha_bar_vec.iter())
+            .enumerate()
+            .map(|(i, (([m0, m1], &t), &b))| {
+                let tweak: Block = bytemuck::cast([i, self.state.exec_counter]);
+                if !b {
+                    // H(t, i|ell) ^ M0
+                    FIXED_KEY_AES.tccr(tweak, t) ^ m0
+                } else {
+                    // H(t, i|ell) ^ M1
+                    FIXED_KEY_AES.tccr(tweak, t) ^ m1
+                }
+            })
+            .collect();
+
+        // Reconstructs GGM tree except `ws[alpha]`.
+        let ggm_tree = GgmTree::new(h);
+        let mut tree = vec![Block::ZERO; 1 << h];
+        ggm_tree.reconstruct(&mut tree, &k, &alpha_bar_vec);
+
+        // Sets `tree[alpha]`, which is `ws[alpha]`.
+        tree[alpha] = tree.iter().fold(sum, |acc, &x| acc ^ x);
+
+        self.state.unchecked_ws.extend_from_slice(&tree);
+        self.state.alphas_and_length.push((alpha, 1 << h));
+
+        self.state.exec_counter += 1;
+
+        Ok(())
+    }
+
+    /// Performs the decomposition and bit-mask steps in check.
+    ///
+    /// See step 7 in Figure 6.
+    ///
+    /// # Arguments
+    ///
+    /// * `x_star` - The message from COT ideal functionality for the receiver. Only the random bits are used.
+    pub fn check_pre(&mut self, x_star: &[bool]) -> Result<CheckFromReceiver, ReceiverError> {
+        if x_star.len() != CSP {
+            return Err(ReceiverError::InvalidLength(
+                "the length of x* should be 128".to_string(),
+            ));
+        }
+
+        let seed = *self.state.hasher.finalize().as_bytes();
+        let mut prg = Prg::from_seed(Block::try_from(&seed[0..16]).unwrap());
+
+        // The sum of all the chi[alpha].
+        let mut sum_chi_alpha = Block::ZERO;
+
+        for (alpha, n) in &self.state.alphas_and_length {
+            let mut chis = vec![Block::ZERO; *n];
+            prg.random_blocks(&mut chis);
+            sum_chi_alpha ^= chis[*alpha];
+            self.state.chis.extend_from_slice(&chis);
+        }
+
+        let x_prime: Vec<bool> = sum_chi_alpha
+            .to_lsb0_vec()
+            .into_iter()
```

Comment:
we can avoid double allocation

```suggestion
            .iter_lsb0()
```